### PR TITLE
Fixes transaction handling within changesets on DB2 on z/OS

### LINKED ISSUE #3299: ChangeSet is not rolled back completely in case of an error for Db2z
## Environment

**Liquibase Version**: 4.15

**Liquibase Integration & Version**: CLI

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: DB2z

**Operating System Type & Version**: Windows 10 Pro

**Infrastructure Type/Provider**:

## Description

Assuming a ChangeSet consists of several steps, among others a CreateTable statement, and an error occurs during execution for any reason, then the entire ChangeSet is rolled back. However, this does not apply to the CreateTable statement and all previous parts of the ChangeSet, these are committed.

If the ChangeSet is executed again after the cause of the error has been corrected, then it fails because the table has already been created.

## Expected/Desired Behavior

The ChangeSet must be rolled back completely, this is not the case for DB2z.

In contrast, it works correctly for DB2luw, among others.

### PR DESCRIPTION
**Pull request by MichaelKern-IVV**
## Impact

- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Erroneous ChangeSets are not always completely rolled back.

Fixes #3299

This is the case because unnecessary and therefore erroneous commits are performed under DB2z. And because liquibase does not honor that DB2z supports DDL in transactions. 

## Things to be aware of

I executed a test script:

`DB2 => connect to db2eha user a61kern 
Aktuelles Kennwort für a61kern eingeben: 

   Datenbankverbindungsinformationen 

 Datenbank-Server                                                = DB2 z/OS 12.1.5 
 SQL-Berechtigungs-ID                                            = A61KERN 
 Aliasname der lokalen Datenbank                                 = DB2EHA 

DB2 => create table test1 (id integer not null primary key, name varchar(20)) 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
DB2 => comment on table test1 is 'Table for auto commit test' 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
DB2 => insert into test1 values (1, 'Test 1') 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
DB2 => select * from test1 

ID          NAME 
----------- -------------------- 
          1 Test 1 

  1 Satz/Sätze ausgewählt. 

DB2 => 
DB2 => commit 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
DB2 => create table test2 (id integer not null primary key, name varchar(20)) 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
DB2 => comment on table test2 is 'Another Table for auto commit test' 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
DB2 => insert into test2 values (1, 'Test 2') 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
DB2 => select * from test2 

ID          NAME 
----------- -------------------- 
          1 Test 2 

  1 Satz/Sätze ausgewählt. 

DB2 => 
DB2 => rollback 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
DB2 => select * from test1 

ID          NAME 
----------- -------------------- 
          1 Test 1 

  1 Satz/Sätze ausgewählt. 

DB2 => 
DB2 => select * from test2 
SQL0204N  "A61KERN.TEST2" ist ein nicht definierter Name.  SQLSTATE=42704 
DB2 => 
DB2 => connect reset 
DB20000I  Der Befehl SQL wurde erfolgreich ausgeführt. 
DB2 => 
`

It can be observed from the outputs that

1. DDL statements are subject to the transaction control and can be rolled back
2. created objects can be used immediately and within the same transaction

## Things to worry about

I only have access to a DB2z 12 and cannot verify for older versions if the behavior is the same.

All the people I have talked to, especially the DB2z administrators in the company, have claimed that the behavior has always been like this and has never been modified.

### CODE REVIEW

File: liquibase-core/src/main/java/liquibase/database/core/Db2zDatabase.java [hunk 1]
```diff
@@ -8,8 +8,8 @@
 import liquibase.util.StringUtil;
 
 public class Db2zDatabase extends AbstractDb2Database {
-	
-	public Db2zDatabase() {
+
+    Db2zDatabase() {
```
Comment (nvoxland) [inline liquibase-core/src/main/java/liquibase/database/core/Db2zDatabase.java:12]: Why did you remove the public from this? I think that will keep this extension from being loaded by the ServiceLoader, because it needs a public no-arg constuctor.

**Comment by nvoxland** [comment]
Thanks for the follow-up on the transaction logic on db2/z, @MichaelKern-IVV.

Unrelated question but while I have you: is the `CURRENT TIMESTAMP` function the correct one for getting the current time on db2/z? Is that something that varies between versions? I've maybe seen some issues since we recently changed that, but I've not gotten a chance to follow through with them yet.

**Comment by MichaelKern-IVV** [comment]
@nvoxland I have now removed the Db2zDatabase construtor completely since it only made the same settings as AbstractDb2Database.

**Comment by MichaelKern-IVV** [comment]
@nvoxland There are three "special registers" to obtain time information:

`DB2 => select current time t, current date d, current timestamp ts from sysibm.sysdummy1 

T        D          TS 
-------- ---------- -------------------------- 
10.08.24 2022-10-07 2022-10-07-10.08.24.144289 

  1 Satz/Sätze ausgewählt.`

As far as I remember (and as long as the IBM documentation goes back) this has always been the case and has never changed.

In short I think `CURRENT TIMESTAMP` is the correct one.

**Comment by nvoxland** [comment]
Thanks for the info and thanks for the extra commits, @MichaelKern-IVV .

**Review by nvoxland — APPROVED** [review]
Code review and test results:

Things to be aware of:
- Only impacts db2 on z

Things to worry about:
- Nothing, contributor answered the questions I had

**Comment by XDelphiGrl** [comment]
@filipelautert, hello! I'm moving this ticket back to Development as the PR is not building for some reason. I'll keep an eye out to see when this comes back to code review. Thanks!

**Review by filipelautert — APPROVED** [review]

**Comment by filipelautert** [comment]
Hi @XDelphiGrl - now it is building fine :)

**Review by XDelphiGrl — APPROVED** [review]
With this PR, Liquibase now honors DDL transactions for the DB2/Z database. 
* No additional testing required.

APPROVED
